### DISCUSSION TITLE
Device Pin/Pattern fallback?

### DISCUSSION THREAD

**Discussion by cking24343** [sergeykomlach/AdvancedBiometricPromptCompat#370] [Q&A] [answered]
Hey all, simple question, forgive me if it's obvious but I'm new to biometrics in general. I was looking over the repo and wondering (mainly because I'm not seeing it), if there is support for allowing the user to fallback to using the device pin / pattern / password for use cases where they cancel out the biometric prompt? 

The only thing I've found so far is a [call](https://github.com/sergeykomlach/AdvancedBiometricPromptCompat/blob/f806e672a4fa111a51318e6a05faf9aba7aecd8b/biometric/src/main/java/dev/skomlach/biometric/compat/impl/BiometricPromptApi28Impl.kt#L264C27-L264C53) where it looks like this has been explicitly disabled? I'm guessing this is intentional and the support for allowing a fallback for a user to cancel biometrics in lieu of device pin / pattern / password is not supported?

**Comment by sergeykomlach** [comment, marked as answer]
Hi.
Sorry for the delay with the answer :(
In general, I didn't expect that fallback would be needed, cause (for me) the main idea of this library - just "pure" biometric authentification.
But, is not a big deal to add CREDENTIALS  (pin/pattern/etc) fallback to the current implementation.
I'll schedule this improvement for the next release

**Reply by sergeykomlach** [reply]
So, I'm on the final stage with implementation for [DeviceCredentials](https://github.com/sergeykomlach/AdvancedBiometricPromptCompat/issues/372) feature.
Stay tuned, new release will be published soon

**Reply by sergeykomlach** [reply]
Fix available in 2.2.15 version.
Please note - due to the OS/vendor fragmentation, the current implementation may behave a bit differently from "native" BiometricPrompt.
Device Credentials will be used if target biometric cannt be used, but not simultaneously with biometric - that's mean OR biometric OR device credentials, but not together
